docs/dotnet/dotnet-kestrel-metrics.md

```diff
@@ -0,0 +1,221 @@
+# Semantic Conventions for Kestrel web server metrics
+
+**Status**: [Experimental][DocumentStatus]
+
+This document defines semantic conventions for Kestrel web server.
+
+<!-- toc -->
+
+- [Metric: `kestrel.active_connections`](#metric-kestrelactive_connections)
+- [Metric: `kestrel.connection.duration`](#metric-kestrelconnectionduration)
+- [Metric: `kestrel.rejected_connections`](#metric-kestrelrejected_connections)
+- [Metric: `kestrel.queued_connections`](#metric-kestrelqueued_connections)
+- [Metric: `kestrel.queued_requests`](#metric-kestrelqueued_requests)
+- [Metric: `kestrel.upgraded_connections`](#metric-kestrelupgraded_connections)
+- [Metric: `kestrel.tls_handshake.duration`](#metric-kestreltls_handshakeduration)
+- [Metric: `kestrel.active_tls_handshakes`](#metric-kestrelactive_tls_handshakes)
+
+<!-- tocstop -->
+
+## Metric: `kestrel.active_connections`
+
+**Status**: [Experimental][DocumentStatus]
+
+This metric is required.
+
+<!-- semconv metric.kestrel.active_connections(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `kestrel.active_connections` | UpDownCounter | `{connection}` | Number of connections that are currently active on the server. [1] |
+
+**[1]:** Corresponding `EventCounter` name is `current-connections`; Meter name is `Microsoft.AspNetCore.Server.Kestrel`
+<!-- endsemconv -->
+
+<!-- semconv metric.kestrel.active_connections(full) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `kestrel.endpoint` | string | TODO [1] | `TODO` | Recommended |
+
+**[1]:** **TODO: why not server.**? What's KEstrel-specific about it**
+<!-- endsemconv -->
+
+## Metric: `kestrel.connection.duration`
+
+**Status**: [Experimental][DocumentStatus]
+
+This metric is required.
+
+<!-- semconv metric.kestrel.connection.duration(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `kestrel.connection.duration` | Histogram | `s` | The duration of connections on the server. [1] |
+
+**[1]:** Meter name is `Microsoft.AspNetCore.Server.Kestrel`
+<!-- endsemconv -->
+
+<!-- semconv metric.kestrel.connection.duration(full) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `exception.type` | string | The type of the exception (its fully-qualified class name, if applicable). The dynamic type of the exception should be preferred over the static type in languages that support it. [1] | `java.net.ConnectException`; `OSError` | Recommended |
```

Comment:
> On Kestrel it would be the exception type name.

Sounds good. 

For error.socket.reason I'd defer to James whether he thought it was useful and justified to add right now in the ASPNET domain and I'd defer to you and others on your team for the HttpClient. I guessing you guys get asked to diagnose issues customers are running into and have at least a basic idea how frequently SocketError was useful in that diagnosis. If it is at least occasionally useful and very straightforward to implement then happy to have it.

Reply:
> error.socket.reason

I'm not proposing to implement this now, but to have an extensible naming scheme that enables adding this kind of lower level error information incrementally.

Reply:
looking into kestrel instrumentation, it seems we just have an exception object. Given it's inside `asp.net.kestrel` namespace, `exception.type` should be fine. 

If we decide to introduce a new error code there, it'd be a behavior-breaking change and could be done in a new major version.